### PR TITLE
Allow deferred connection initialization

### PR DESCRIPTION
**Pull request by mors741**
Implements https://github.com/lettuce-io/lettuce-core/issues/1129

### CODE REVIEW

**Comment by zhangyongxianggithub** [comment]
ewrwrweqeewq

File: src/main/java/io/lettuce/core/AbstractRedisClient.java [hunk 1]
```diff
@@ -320,23 +320,31 @@ private void initializeChannelAsync0(ConnectionBuilder connectionBuilder, Comple
...
 
             if (!future.isSuccess()) {
-
                 logger.debug("Connecting to Redis at {}: {}", redisAddress, future.cause());
-                connectionBuilder.endpoint().initialState();
-                channelReadyFuture.completeExceptionally(future.cause());
-                return;
+                if (clientOptions.isAllowDeferredConnection() && clientOptions.isAutoReconnect()) {
+                    logger.debug("Connection deferred - will try to reconnect in background");
+                    future.channel().pipeline().fireUserEventTriggered(new ConnectionEvents.Deferred());
```
Comment (zhangyongxianggithub) [inline src/main/java/io/lettuce/core/AbstractRedisClient.java]: weqewqeqwfrfewqewqewqewq

File: src/main/java/io/lettuce/core/AbstractRedisClient.java [hunk 2]
```diff
@@ -320,23 +320,31 @@ private void initializeChannelAsync0(ConnectionBuilder connectionBuilder, Comple
             }
         });
 
-        connectFuture.addListener(future -> {
+        connectFuture.addListener((ChannelFutureListener) future -> {
 
             if (!future.isSuccess()) {
```
Comment (zhangyongxianggithub) [inline src/main/java/io/lettuce/core/AbstractRedisClient.java]: t666666

**Comment by zhangyongxianggithub** [comment]
wqewqeeqwewqewq

**Comment by mp911de** [comment]
Closing as per comment in the associated ticket #1129. Thanks for your effort, but we've decided to not support initially failing connections.

**Pull request closed without merging**